File: project_euler_unified/problem_610/solution.py
"""
Problem 610: Roman Numerals II
Count valid representations in modified Roman numeral system.
"""

def roman_to_int(s):
    """Convert Roman numeral string to integer."""
    values = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
    total = 0
    for i in range(len(s)):
        if i + 1 < len(s) and values[s[i]] < values[s[i+1]]:
            total -= values[s[i]]
        else:
            total += values[s[i]]
    return total

def int_to_roman(n):
    """Convert integer to standard Roman numeral."""
    vals = [(1000,'M'),(900,'CM'),(500,'D'),(400,'CD'),
            (100,'C'),(90,'XC'),(50,'L'),(40,'XL'),
            (10,'X'),(9,'IX'),(5,'V'),(4,'IV'),(1,'I')]
    result = ''
    for val, sym in vals:
        while n >= val:
            result += sym
            n -= val
    return result

# Verify
assert roman_to_int("XIV") == 14
assert roman_to_int("MCMXCIV") == 1994
assert int_to_roman(1994) == "MCMXCIV"

# Count lengths of Roman numerals
lengths = {}
for n in range(1, 4000):
    r = int_to_roman(n)
    lengths[n] = len(r)

print(f"Longest Roman numeral under 4000: {max(lengths.values())} chars")
longest_n = max(lengths, key=lengths.get)
print(f"  Number: {longest_n} = {int_to_roman(longest_n)}")

File: project_euler_unified/problem_610/solution.cpp
#include <bits/stdc++.h>
using namespace std;

string int_to_roman(int n) {
    int vals[] = {1000,900,500,400,100,90,50,40,10,9,5,4,1};
    string syms[] = {"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"};
    string result;
    for (int i = 0; i < 13; i++)
        while (n >= vals[i]) { result += syms[i]; n -= vals[i]; }
    return result;
}

int main() {
    int max_len = 0;
    for (int n = 1; n < 4000; n++) {
        int len = int_to_roman(n).size();
        max_len = max(max_len, len);
    }
    cout << "Max Roman numeral length: " << max_len << endl;
    return 0;
}
